quinn-proto/src/connection/spaces.rs

```diff
@@ -476,6 +479,88 @@ impl PendingAcks {
     }
 }
 
+/// Helper for mitigating [optimistic ACK attacks]
+///
+/// A malicious peer could prompt the local application to begin a large data transfer, and then
+/// send ACKs without first waiting for data to be received. This could defeat congestion control,
+/// allowing the connection to consume disproportionate resources. We therefore occasionally skip
+/// packet numbers, and classify any ACK referencing a skipped packet number as a transport error.
+///
+/// Skipped packet numbers occur only in the application data space (where costly transfers might
+/// take place) and are distributed exponentially to reflect the reduced likelihood and impact of
+/// bad behavior from a peer that has been well-behaved for an extended period.
+///
+/// ACKs for packet numbers that have not yet been allocated are also a transport error, but an
+/// attacker with knowledge of the congestion control algorithm in use could time falsified ACKs to
+/// arrive after the packets they reference are sent.
+///
+/// [optimistic ACK attacks]: https://www.rfc-editor.org/rfc/rfc9000.html#name-optimistic-ack-attack
+pub(super) struct OptimisticAckDetector {
```

Comment:
Naming considerations: so the name here describes the purpose (detecting opportunistic ACKs) well but doesn't really describe the mechanism (skipping packet numbers). In turn, this suggests making the method names fairly verbose (adding the `_packet_number` suffix). Do you think that's the right trade-off?

Reply:
Good thought; renamed to `PacketNumberFilter`.